src/mysql.erl

```diff
@@ -155,37 +155,47 @@ start_link(Options) ->
         _ -> ok
     end,
     Ret.
-
+ 
+%% @see stop/2.
 -spec stop(Conn) -> ok
     when Conn :: connection().
 stop(Conn) ->
     stop(Conn, infinity).
 
+%% @doc Stops a connection process and closes the connection. The
+%% process calling `stop' will be blocked until the connection
+%% process stops or the given timeout expires.
+%%
+%% If the connection is not stopped within the given timeout,
+%% an exit exception is raised with reason `timeout'.
+%%
+%% If the connection process exits with any other reason than `normal',
+%% an exit exception is raised with that reason.
 -spec stop(Conn, Timeout) -> ok
     when Conn :: connection(),
          Timeout :: timeout().
 stop(Conn, Timeout) ->
-    case erlang:function_exported(gen_server, stop, 3) of
-        true -> gen_server:stop(Conn, normal, Timeout);
-	false -> Monitor=monitor(process, Conn),
-                 exit(Conn, normal),
-                 receive
-                     {'DOWN', Monitor, process, Conn, normal} ->
-                         ok;
-                     {'DOWN', Monitor, process, Conn, Reason} ->
-                         exit(Reason)
-                 after Timeout ->
-                     demonitor(Monitor, [flush]),
-                     exit(timeout)
-                 end
+    stop(erlang:function_exported(gen_server, stop, 3), Conn, Timeout).
```

Comment:
Maybe it's a matter of personal taste, but I was anticipating something like

```Erlang
    case erlang:function_exported(gen_server, stop, 3) of
        true  -> gen_server:stop(Conn, normal, Timeout);           %% OTP >= 18
        false -> backported_gen_server_stop(Conn, normal, Timeout) %% OTP < 18
    end
```

If you don't like my suggestion, I'm not going to be picky about it.

Reply:
Nah, either one is fine with me :) Did it.

This should be more or less the final version of it. In the tests, the `exit(Pid, normal)` calls have been swapped out for `mysql:stop`, which should provide a lot of coverage for it already. If you want, I can add some more tests dedicated to stopping to it, though.

I followed `gen_server:stop` through the OTP code, and found it is using `gen:stop`, which is using `proc_lib:stop`, which is using `sys:terminate`, all of which appeared with OTP 18, so that kinda explains why it didn't appear earlier. Following from there is harder as it starts sending messages around, and I was running out of time, so I stopped there ;) I guess the backported `gen_server;stop` is not a full backport, but as close as I could make it.

What `proc_lib:stop` does when it hits the timeout is kill the process. My former iterations of the backported function didn't do that, and I put that in with my last commit. What this means is that it will take down the process linked to the connection process, also, unless it is trapping exits. While I'm not sure that this is a very great thing, it is what `gen_server:stop` ultimately does, so the backport needs to do it as well in order to avoid inconsistent behavior between pre- and post-18 OTP releases.

Reply:
Oh, huh, the tests fail. Looks like I messed something up with the commit :P

Reply:
Ah, better :) Strange, how could that have ever worked locally :P

Reply:
I can't see what you did here since you force-pushed, but I don't need to know...

Reply:
> What `proc_lib:stop` does when it hits the timeout is kill the process. My former iterations of the backported function didn't do that, and I put that in with my last commit.

Good investigation! Kill is good. It's what a supervisor does when a child doesn't terminate within the specified time. A supervisor, poolboy and similar would trap_exit, so no worries for those.